### ISSUE TITLE
Using Weld SE 2.4.4.Final (current stable version) breaks cucumber-weld v1.2.5

### ISSUE THREAD

**Issue by novityFS** [cucumber/cucumber-jvm#1165]
## Summary

Using `Weld SE 2.4.4.Final` (current stable version) breaks `cucumber-weld v1.2.5`.

## Expected Behavior

Given a simple feature file
And a JUnit class running with Cucumber.class runner
And using cucumber-weld version v1.2.5 with current stable Weld SE version (2.4.4.Final)
When running the test
Then cucumber should run the test

## Current Behavior

Given a feature file with more than one scenario
And a JUnit class running with Cucumber.class runner
And using cucumber-weld version v1.2.5 with current stable Weld SE version (2.4.4.Final)
When running the test
Then an IllegalStateException is thrown

```
Running cucumber.runtime.java.weld.RunCukesTest
Jul 05, 2017 11:13:46 PM org.jboss.weld.bootstrap.WeldStartup <clinit>
INFO: WELD-000900: 2.4.4 (Final)
Jul 05, 2017 11:13:46 PM org.jboss.weld.config.WeldConfiguration readObsoleteFileProperties
WARN: WELD-001904: Unsupported configuration key found and ignored: enabled in {enabled=true}
Jul 05, 2017 11:13:46 PM org.jboss.weld.bootstrap.WeldStartup startContainer
INFO: WELD-000101: Transactional services not available. Injection of @Inject UserTransaction not available. Transactional observers will be invoked synchronously.
Jul 05, 2017 11:13:46 PM org.jboss.weld.environment.se.WeldContainer fireContainerInitializedEvent
INFO: WELD-ENV-002003: Weld SE container 5e6f6274-9362-402d-983b-6c36cee655f6 initialized
Jul 05, 2017 11:13:46 PM org.jboss.weld.environment.se.WeldContainer shutdown
INFO: WELD-ENV-002001: Weld SE container 5e6f6274-9362-402d-983b-6c36cee655f6 shut down
Jul 05, 2017 11:13:46 PM org.jboss.weld.config.WeldConfiguration readObsoleteFileProperties
WARN: WELD-001904: Unsupported configuration key found and ignored: enabled in {enabled=true}
Jul 05, 2017 11:13:46 PM org.jboss.weld.bootstrap.WeldStartup startContainer
INFO: WELD-000101: Transactional services not available. Injection of @Inject UserTransaction not available. Transactional observers will be invoked synchronously.
Jul 05, 2017 11:13:47 PM org.jboss.weld.environment.se.WeldContainer fireContainerInitializedEvent
INFO: WELD-ENV-002003: Weld SE container a1a97db5-59a8-41bc-ad4b-1bcbff6b35a6 initialized
Jul 05, 2017 11:13:47 PM org.jboss.weld.environment.se.WeldContainer shutdown
INFO: WELD-ENV-002001: Weld SE container a1a97db5-59a8-41bc-ad4b-1bcbff6b35a6 shut down

2 Scenarios ([32m2 passed[0m)
4 Steps ([32m4 passed[0m)
0m0,093s

Tests run: 6, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 1.257 sec <<< FAILURE! - in cucumber.runtime.java.weld.RunCukesTest
Feature: Cukes  Time elapsed: 0.011 sec  <<< ERROR!
java.lang.IllegalStateException: WELD-ENV-002002: Weld SE container 5e6f6274-9362-402d-983b-6c36cee655f6 was already shut down
```

Given the cucumber source code
When running the tests with `maven test`
Then an IllegalStateException is thrown in the `WeldFactoryTest`

```
Running cucumber.runtime.java.weld.WeldFactoryTest
Jul 05, 2017 11:13:47 PM org.jboss.weld.config.WeldConfiguration readObsoleteFileProperties
WARN: WELD-001904: Unsupported configuration key found and ignored: enabled in {enabled=true}
Jul 05, 2017 11:13:47 PM org.jboss.weld.bootstrap.WeldStartup startContainer
INFO: WELD-000101: Transactional services not available. Injection of @Inject UserTransaction not available. Transactional observers will be invoked synchronously.
Jul 05, 2017 11:13:47 PM org.jboss.weld.environment.se.WeldContainer fireContainerInitializedEvent
INFO: WELD-ENV-002003: Weld SE container 97ff73cb-0ba8-44ac-ad40-b9411ff58b9c initialized
Jul 05, 2017 11:13:47 PM org.jboss.weld.environment.se.WeldContainer shutdown
INFO: WELD-ENV-002001: Weld SE container 97ff73cb-0ba8-44ac-ad40-b9411ff58b9c shut down
Jul 05, 2017 11:13:47 PM org.jboss.weld.config.WeldConfiguration readObsoleteFileProperties
WARN: WELD-001904: Unsupported configuration key found and ignored: enabled in {enabled=true}
Jul 05, 2017 11:13:47 PM org.jboss.weld.bootstrap.WeldStartup startContainer
INFO: WELD-000101: Transactional services not available. Injection of @Inject UserTransaction not available. Transactional observers will be invoked synchronously.
Jul 05, 2017 11:13:47 PM org.jboss.weld.environment.se.WeldContainer fireContainerInitializedEvent
INFO: WELD-ENV-002003: Weld SE container ffe0c1a0-fb6b-4e75-9b9d-5e7998bcbf09 initialized
Jul 05, 2017 11:13:47 PM org.jboss.weld.environment.se.WeldContainer shutdown
INFO: WELD-ENV-002001: Weld SE container ffe0c1a0-fb6b-4e75-9b9d-5e7998bcbf09 shut down
Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.084 sec <<< FAILURE! - in cucumber.runtime.java.weld.WeldFactoryTest
shouldGiveUsNewInstancesForEachScenario(cucumber.runtime.java.weld.WeldFactoryTest)  Time elapsed: 0.083 sec  <<< ERROR!
java.lang.IllegalStateException: WELD-ENV-002002: Weld SE container 97ff73cb-0ba8-44ac-ad40-b9411ff58b9c was already shut down
        at cucumber.runtime.java.weld.WeldFactoryTest.shouldGiveUsNewInstancesForEachScenario(WeldFactoryTest.java:23)


Results :

Tests in error:
  WELD-ENV-002002: Weld SE container 5e6f6274-9362-402d-983b-6c36cee655f6 was already shut down
  WeldFactoryTest.shouldGiveUsNewInstancesForEachScenario:23 » IllegalState WELD...
```

## Possible Solution

I've forked the cucumber-jvm project into https://github.com/novityFS/cucumber-jvm and created a fix in the branch `v1.2.5-fix`. As soon as this issue is accepted, I can provide a pull request, If you like.

## Steps to Reproduce (for bugs)

1. Checkout the branch `v1.2.5-issue` from repository https://github.com/novityFS/cucumber-jvm
2. Run tests with `mvn clean test`
3. See errors shown above

The changes made to the original version are:
* Bumped the Weld SE version to current stable version
* Duplicated the first scenario in the cukes.feature file

## Context & Motivation

I'm using cucumber for (integration) testing a JEE application using CDI as DI mechanism. I've run into this problem by starting a new maven project, today, using the current version of Weld SE and got the IllegalStateException.

First I thought that this was a problem with my test. By simplifying, using your own tests and some debugging I've come to the conclusion, that this might be a bug.

## Your Environment

* Version used: v1.2.5
* Operating System and version: CentOs (CI System) and Windows 10 (Developer System)

**Comment by mpkorstanje** [comment]
Looks like a valid issue. The fix also looks good at a glance. Please
provide a PR. I'll try to merge it later this week.

Looks like we violated the "prefer composition over inheritance" principle
but I do need to spend a little time with Weld to confirm I understand the
problem.

**Comment by novityFS** [comment]
Ok, I've submitted Pull Request #1166.

Please keep in mind that this fix is for the v1.2.5 version of cucumber and currently not checked with the 2.x version

**Comment by mpkorstanje** [comment]
The problem was caused by a change in Weld and an improper implementation in cucumber. When destroying the backend worlds cucumber would attempt to shut down all cdi instances that were created by the builder. Even those that had been shut down in a previous test. When attempting to shut down a container that was not running Weld 2.3.5.Final would ignore this while Weld SE 2.4.4.Final throws an exception.

The fix for this will be released in 2.0.0.
